fix(axis): fix axis label may have inappropriate precision or take too much unexpected space when `alignTicks` is enabled

plainheart opened the pull request:
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others



### What does this PR do?

This PR is to fix the following issues when `alignTicks` is enabled:

1. Fix axis label may have inappropriate precision.
2. Fix axis label takes too much unexpected space.

### Fixed issues

- Fix #21552
- Fix #21116
- Fix #19967 
- Fix #18063 
- Fix #17607

## Comparison

| Before | After |
| :----: | :----: |
| <img width="1925" height="1175" alt="image" src="https://github.com/user-attachments/assets/8ef6b421-562d-49aa-90b0-a9724a44744b" /> |<img width="1925" height="1182" alt="image" src="https://github.com/user-attachments/assets/d525210e-73b4-4e8a-919c-191da2dd4381" /> |
| <img width="1922" height="1181" alt="image" src="https://github.com/user-attachments/assets/83febb81-0c58-42a9-a79b-ae8869414f02" /> | <img width="1922" height="1181" alt="image" src="https://github.com/user-attachments/assets/a872aacc-6e6d-41ba-b31c-8dcdae1ed4bd" /> |
| <img width="1917" height="1127" alt="image" src="https://github.com/user-attachments/assets/f79baca0-5906-47a3-adc3-ba4b78c675da" /> | <img width="1920" height="1132" alt="image" src="https://github.com/user-attachments/assets/50e8638c-8b34-4755-850c-719937eecd8b" /> |
| <img width="1845" height="990" alt="image" src="https://github.com/user-attachments/assets/63d99d81-3137-46a2-9a48-544177e7faaa" /> | <img width="1847" height="995" alt="image" src="https://github.com/user-attachments/assets/ee6e60c4-0724-4dbb-b215-3ca305074ef8" /> |


## Document Info

One of the following should be checked.

- [x] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx



## Misc

### Security Checking

- [ ] This PR uses security-sensitive Web APIs.

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

See the 6th/7th/8th test cases in `test/axis-align-ticks.html`

### Merging options

- [ ] Please squash the commits into a single one when merging.

### Other information

plainheart committed:
fix(axis): fix axis label may have inappropriate precision or take too much unexpected space when `alignTicks` is enabled

nuteu commented:
@plainheart  请问这个mr后续会回合到 5.x 版本吗

100pah committed:
test(dataZoom): fix test case -- accidentally broken in v6. And add more case for dataZoom edge cases.

100pah committed:
timeScale: Add the missing rounding (broken from a previous version) to time scale.

100pah committed:
fix&feat: Change and clarify the rounding error and auto-precision utils and solutions.

100pah committed:
fix(axisTick&dataZoom): (1) Apply a better auto-precision method. (2) Make the rounding result consistent between dataZoom calculated window and specified axis `determinedMin/Max`. (3) Fix unexpected behaviors when dataZoom controls axes with `alignTicks: true` - previous they are not precisely aligned and the ticks jump significantly due to inappropriate rounding when dataZoom dragging.

100pah committed:
fix(alignTicks): Change alignTick strategy: (1) Previously some series data may be out of the calculated extent and can not be displayed. (2) Previously the precision is incorrect for small float number (fixed at 10 rather than based on the magnitude of the value). (3) Make the tick precision more acceptable when min/max of axis is fixed, and remove console warning, because whey can be specified when dataZoom dragging. (4) Clarify the related code for LogScale.

100pah committed:
feat(alignTicks): (1) Fix LogScale precision. (2) Tweak align ticks layout. (3) Remove unreasonable clamp in Interval calcNiceExtent, and clarify the definition of `_niceExtent`.

100pah commented:
Additional modifications:


## Comprehensively fixed bad cases of "alignTicks" and "precision"

### Series data disappear in some cases

(previously => now)
<img width="1069" height="230" alt="image" src="https://github.com/user-attachments/assets/52b4857e-8c7a-4c6e-a318-d7ff72af4c3e" />

<img width="1081" height="203" alt="image" src="https://github.com/user-attachments/assets/28179d7c-e826-48fd-a3cc-537a8ff1295e" />


### Rounding error occur on min/max/ticks

(previously => now)
<img width="1069" height="256" alt="image" src="https://github.com/user-attachments/assets/b9d0817a-0b45-45f1-9f6c-dced6b035c0d" />

<img width="1072" height="146" alt="image" src="https://github.com/user-attachments/assets/730b44a7-3042-4b2d-9655-31e2c5441fbc" />

### Better auto precision

(previously => now)
<img width="1084" height="243" alt="image" src="https://github.com/user-attachments/assets/db54de57-6513-489a-8ffa-fe29860816f7" />
<img width="1083" height="252" alt="image" src="https://github.com/user-attachments/assets/1102c7a2-ba73-487f-a194-a57c418b1b32" />


### Incorrect ticks for small series data

(previously => now)
<img width="1071" height="250" alt="image" src="https://github.com/user-attachments/assets/b80dd29f-0148-48c3-8cf5-ce5eababe793" />
<img width="1075" height="240" alt="image" src="https://github.com/user-attachments/assets/ceb228ed-5937-4110-9932-44331d57587d" />



### Lay out series data to the center of Y

(previously => now)
<img width="1067" height="300" alt="image" src="https://github.com/user-attachments/assets/31e349e6-3f31-40a9-929d-27aedd249fcd" />


### Bad cases when alignTo axis min or max is set

(previously => now)
<img width="1088" height="242" alt="image" src="https://github.com/user-attachments/assets/d0f581dd-7ede-4ea2-af3c-efedf9215af7" />

### Reasonable effect when dataZoom controls all yAxis

(previously => now)
<img width="1110" height="259" alt="image" src="https://github.com/user-attachments/assets/4b60c4de-8621-45b0-9bd8-d8d2cb4ee494" />

<img width="1094" height="240" alt="image" src="https://github.com/user-attachments/assets/5bd24f13-bd7b-4df8-ae55-afb054b01323" />

<img width="1112" height="246" alt="image" src="https://github.com/user-attachments/assets/6c8c5725-9544-46a1-85c8-4727d1cf9da0" />


### Radar bad case

(previously => now)
<img width="1074" height="294" alt="image" src="https://github.com/user-attachments/assets/e4e1228a-da70-4f04-a8ae-e4fdd67a8af8" />

## Fix incorrect align ticks in some cases

`test/axis-align-ticks-random.html`





## Fix dataZoom

+ Respect to axis ends and better auto-precision
+ Fix dataZoom hovering bug - previously it did not always show labels when hovering.

(previously => now)
<img width="1073" height="190" alt="Image" src="https://github.com/user-attachments/assets/41a44929-4721-4fbb-ab06-f0e325296ecf" />



## Fix toolbox theme and update

Previously, toolbox settings in theme can cause it always show, since toolbox will display if `{}` exists, e.g., `toolbox: {feature: {some: {}}`. Now enable `{show: boolean}` without breaking changes. And enable toolbox features change by `setOption`.


## Fix progressive rendering

See test cases:
test/stream-basic1.html
test/stream-basic2.html
test/stream-basic3.html
test/stream-basic4.html

This fix is mainly implemented in https://github.com/ecomfe/zrender/pull/1151



## Fix overflow on axis and Cartesian and enable bandWidth

Previously,
1. On "category" axis with `boundaryGap: false`, series like bar/candlestick/boxplot overflow Cartesian area , thereby no proper clipping, and may cover axis labels, causing visual artifacts.
2. On "time" axis, bar overflow has been fixed but need to refactor to "register" style for code size; other similar series is not covered.
3. On "value"/"log"/"time" axis, overflow is not fixed.
4. Candlestick edge items (min or max items) may not be displayed.

Now, this issue is systematically resolved. 
- A new ec option `xxxAxis.containShape: boolean` is introduced, `true` by default.
- Axis is auto expanded to contain series shape, typically by "half" `bandWidth` at the start and end.
- `bandWidth` calculation is fixed for `axisPointer: shadow` in this cases.

(previously => now)
<img width="693" height="221" alt="Image" src="https://github.com/user-attachments/assets/d2c55e76-9d9c-4aff-965b-dedbb39f1a77" />

<img width="691" height="179" alt="Image" src="https://github.com/user-attachments/assets/afe12e53-dc00-4ff1-91af-23ca176c841e" />

<img width="715" height="168" alt="Image" src="https://github.com/user-attachments/assets/bad1d288-d2dd-423c-ac2b-e1f9d37996a5" />

<img width="680" height="220" alt="Image" src="https://github.com/user-attachments/assets/0f127440-904a-4257-aece-e7fdc915f9e8" />

category Axis with `boundaryGap: false`:

(previously => now)
<img width="1008" height="209" alt="Image" src="https://github.com/user-attachments/assets/4757e427-f2a5-456e-80f8-fa952cc22ebc" />

dataZoom clip:

(previously => now)
<img width="1052" height="259" alt="Image" src="https://github.com/user-attachments/assets/4680306e-49ee-4258-86ae-0a8e2d8e13d6" />

bandWidth incorrect:

(previously => now)
<img width="710" height="241" alt="Image" src="https://github.com/user-attachments/assets/f34e291f-5621-4c82-9065-d7b18eb7acdd" />



## Auto onZero change

(previously => now)
<img width="1062" height="221" alt="Image" src="https://github.com/user-attachments/assets/df1eda6f-9370-4ad4-8ab0-baba452fb7ee" />

Introduced `onZero: 'auto'` (by default).



## Enable LogScale (axis.type: 'log') to automatically filter out non-positive values

(previously => now)
<img width="687" height="243" alt="Image" src="https://github.com/user-attachments/assets/8b324eee-9ef3-4393-83c6-98ecad8c08e8" />

Related issues: #13154 #17322

## Enhance pie laying out on Cartesian

Previously, if pie was laid out on a Cartesian, its axis extent did not union pie position. Users needed to manually add some hidden series to expand the axis extent. Now auto-union pie center to axis extent is supported.

(previously => now)
<img width="681" height="179" alt="Image" src="https://github.com/user-attachments/assets/427686f4-73ab-4203-89dd-488131742c31" />

See also test case `test/pie-coordinate-system.html`


## Fix hoverLayer rendering artifact bug

Fix https://github.com/apache/echarts/issue/21233
This is mainly fixed in https://github.com/ecomfe/zrender/pull/1151


## Fix geo roaming visual artifact

(previously => now)
<img width="1079" height="276" alt="Image" src="https://github.com/user-attachments/assets/6f368441-1ba1-49ca-95b4-3090feb08fda" />

<img width="446" height="433" alt="Image" src="https://github.com/user-attachments/assets/5da9e81e-6d21-4e6e-aaa1-a070f7a37245" />

This is caused by failing to synchronize roaming variables (center, zoom and corresponding affine matrix) while animation is applied on coord sys view (such as map).


## Support scatter/effectScatter clipping on geo coord sys 

(previously => now)
<img width="1076" height="255" alt="Image" src="https://github.com/user-attachments/assets/2d98d44b-5a31-48a6-8b9f-30babf2274cd" />


## Refactor

Corresponding refactors are performed to the changes above, mainly including:
- Axis scale hierarchy implementation is changed to a pattern like decorator, which covers the abstraction of numeric transform composition (axis break, logarithm, ordinal sort, ...), and enables more transform to be implemented in future (like `asinh`)
- Introduce axis scale extent kind `SCALE_EXTENT_KIND_EFFECTIVE` and `SCALE_EXTENT_KIND_MAPPING` for `containShape`.
- Introduce axis statistics stage to collect `<axis, series>` for axis extent union and series like bar/boxplot/..., and bandWidth calculation on numeric `Scale` and reuse.
- Introduce axis bandWidth calculation and reuse for both category axis and numeric axes.
- Clarify axis extent calculation process (including union from series data and adopt ec options). For historical reason, it is hard to maintain and error-prone.
- Clarify axis nice ticks implementation. For historical reason, it is hard to maintain and error-prone.


## Some related test cases

+ test/toolbox-custom.html 
+ test/candlestick-large3.html 
+ test/bar-stack.html 
+ test/bar-datazoom-filtermode-none-time-axis.html 
+ test/bar-race2.html 
+ test/boxplot-multi.html 
+ test/bar-polar-multi-series-radial.html 
+ test/bar-overflow-time-plot.html  
+ test/bar-overflow-plot2.html 
+ test/dataZoom-action.html 
+ test/candlestick-case.html 
+ test/bar-polar-multi-series.html 
+ test/dataZoomHighPrecision.html 
+ test/hover-style.html 
+ test/stream-basic1.html 
+ test/stream-basic2.html 
+ test/stream-basic3.html 
+ test/stream-basic4.html 
+ test/stream-basic5.html 
+ test/hover-layer.html 
+ test/axis-interval2.html 
+ test/axis-extreme2.html 
+ test/clip.html


## Some related issues

+ https://github.com/apache/echarts/pull/20718
+ https://github.com/apache/echarts/issues/21552 
+ https://github.com/apache/echarts/issues/18688
+ https://github.com/apache/echarts/issues/19972
+ https://github.com/apache/echarts/issues/20503
+ https://github.com/apache/echarts/issues/20421
+ https://github.com/apache/echarts/issues/12720
+ https://github.com/apache/echarts/issues/17858
+ https://github.com/apache/echarts/issues/13321
+ https://github.com/apache/echarts/issues/11240
+ https://github.com/apache/echarts/issues/10374
+ https://github.com/apache/echarts/pull/18202

100pah committed:
refactor(scale): For readability and maintainability (1) Migrate `calcNiceTicks` and `calcNiceExtent` from Scale class override member functions to plain functions, similar to `axisAlignTicks`. Previously it's hard to modify and error-prone. (2) Remove unnecessary override on Scale class hierarchy and limit override usage, which is difficult to understand and error-prone. (3) Simplify the inheritance - shift `LogScale` and `TimeScale` inheritance from `IntervalScale` to `Scale`. (4) Clarify the impl of `IntervalScale` - uniform the parameters setting entry for both "nice ticks" and "align ticks".

100pah committed:
refactor&fix: (1) Unify series data union code - remove union code from Scale; unify union entry and scaleRawExtentInfo creator and cache to avoid error-prone impl; simplify the impl of coord sys call extent uinon. (2) Fix LogScale precision bug introduced by previous commits.

100pah committed:
fix(logScale): (1) Thoroughly resolve a long-standing issue of non-positive data on LogScale - exclude non-positive series data items when calculate dataExtent on LogScale. (2) Include `Infinite`  into `connectNulls` handling on line series; the `Infinite` value may be generated by `log(0)` and previously the corresponding effect in unpredictable on line series (sometimes display as connected but sometimes not).

100pah committed:
fix(pie-on-cartesian): Previously when pie is located on Cartesian, axes extent can not union pie center automatically. Commit 18a23a87585cc8a1575b63dcc4060af5007330fc has supported it. This commit updates test cases.

100pah committed:
fix(dataZoom): (1) Fix dataZoom dragging cursor style (use 'drag' and 'dragging'). (2) Fix dataZoom bug that data info disappear when drag release on the area that data info display should be triggered.

100pah committed:
refactor&feature:
(1) Uniform series data union logic; previous they are implemented in several inconsistent ways and error-prone. And also strict the extent calculation with uniformed utilities.
(2) Refactor Scale - uniform the implementation of the numeric transformations (such as axis breaks, logarithm, ordinal related handling) with a decorator pattern and decouple them from Scale class hierarchy. This brings consistent handling for details and enables further extension.
(3) Refactor Scale - decouple "axis nice" logic from Scale and make the "axis nice" and "axis align" implemented in the same code style. This removes override (hard to read and maintain), and uniforms the parameters of IntervalScale (previous its parameters can be generated by itself or outside, which is confusing), and reduce scale-type-specified handling (e.g., if (isLogScale) ...) in calculation.
(4) Enable "value" axis and "log" axis to render "bar" and "pictorialBar" series without overflowing the axis. Previously only "time" axis enables that on only "bar" series. Introduce "SCALE_EXTENT_KIND_MAPPING" to cover this generalized behavior and use strict "clip" for bar series without extra margin (a previous compromise). And clarify `barGrid` code accordingly, which removes duplicated time-consuming calculation in "bandWidth" calculation. Close #19972. Close #17858. Close #12720. Close #13321. Close #20718. Close #20421. Close #20503.

100pah committed:
test(visual): Update markers.

100pah committed:
Merge branch 'release' into pr/plainheart_fix/alignTicks-precision

# Conflicts:
#	src/chart/bar/BarView.ts
#	src/chart/treemap/treemapLayout.ts

100pah committed:
refactor: Remove the default value of number round due to its unreasonable and error-prone for small float number.

100pah committed:
fix: (1) Fix dataZoom AxisProxy can not be cleared when dataZoom option changed. (2) Fix onZero on double value axis and dataZoom is applied on a base axis - onZero should be disabled by default.

100pah committed:
fix(dataZoom): Do not display values outside of effective extent.

100pah committed:
chore: Tweak the usage of isFinite.

100pah committed:
feature&fix(axis):
(1) feature: Enable uniform bandWidth calculation in numeric axis (e.g., for tooltip shadow); it previously only applicable to category axis, but buggy in numeric axis with bar series. And enable the clip of tooltip shadow.
(2) refactor: Introduce a dedicated workflow phase for series aggregation and data statistics computation on a single axis, allowting the results to be reused across multiple features.
(3) fix: Fix duplicate ticks in TimeScale and customValues, which cause jitter of splitArea.
(4) fix: Fix category showMin/MaxLabel handling when step > 1 and showMin/MaxLabel: false
(5) chore: Tweak bad effects introduced by the previous implementation of SCALE_EXTENT_KIND_MAPPING.
(6) chore: Clean some code.

100pah committed:
fix: Fix inappropriate impl introduced by the previous commits.

100pah committed:
fix&feature:
(1) feature: Add option boxplot.clip
(2) feature: Enable boxplot and candlestick containShape for "value"/"time"/"log" axis, enable proper shadow axisPointer. (candlestick-case.html, boxplot-multi.html)
(3) fix: Fix candlestick ends shapes can not be displayed on "value"/"time" axis. (candlestick-case.html)
(4) fix: candlestick shape width is inappropriate when datazoom filterMode is 'none'/'empty'.  (candlestick-case.html)
(5) fix: In polar coordinate system, support "value"/"time" axis as angle/radius axis, enable propere bandWidth for bar series, enable proper shadow axisPointer. (bar-polar-multi-series-radial.html, bar-polar-multi-series.html)

100pah committed:
fix: Temporarily fix incorrect stack base dimension selection when both axes are value axes.

100pah committed:
chore: Sync the modification of #21448 to release.

100pah committed:
fix: (1) Clarify and uniform terminology and add comments to explain EC_CYCLE. (2) Fix dataZoom in `appendData`, introduced by recent commits.

100pah committed:
fix(appendData): Fix that the dataZoom inside is disabled when appendData is executed. And clarify the usage of appendData in comments.

100pah committed:
tweak: Clarity the previous implements of axis statistics.

100pah committed:
Merge branch 'release' into PR/plainheart_fix/alignTicks-precision

100pah committed:
fix(toolbox): Fix that toolbox theme cause corresponding icons are always displayed even if not required. See #21176 .

100pah committed:
fix(toolbox): Simplify toolbox and fix that toolbox throw error when remove DataZoom feature.

100pah committed:
fix: fix previous commit.

100pah committed:
fix: (1) Previously hoverLayerThreshold modification does not work. (2) Restrict the triggering of hoverLayer to only canvas renderer.

100pah committed:
fix(axisPointer&tooltip): (1) axisPointer and tooltip should be able to update when mousewheel, since dataZoomInside can modify views on mousewheel, and cause highlighted element to be not able to restore. (2) Fix axisPointer highlighted item can not restore due to outdated dataIndexIndex.

100pah committed:
test: add hover layer test.

100pah committed:
fix(progressive): (1) Fix that candlestick large data progressive rendering was broken by Doji feature. (2) Fix bar large and progressive might throw NPE. (3) Follow the fixing of progressive rendering in zrender#1151 (3) Fix parallels did not clear previous rendered content when switching between progressive and normal mode. (4) Fix that chart.on('finished') event might not be called in some edge case (when multiple iterations are executed in one frame at the end).

100pah committed:
fix: Fix min/max label show in category axis. Previously, axisLabel.showMinLabel/showMaxLabel did not control label and ticks on axis edges, but controls the filtered labels by axisLabel.interval settings. This impl did not align with the intent of that option.

100pah committed:
Fix bugs introduced by preceding commit.

100pah committed:
Fix category axis min/max ticks/labels show hide when interval > 0. Remove unnecessary code. Fix relevant TS.

100pah committed:
Supplement tests and fix boxplot (broken by preceding commits)

100pah committed:
fix: Simplify and clarify the code about axis ticks regarding ordinal scale sorting. Fix axis label and sorting bug introduced by preceding commits.

100pah committed:
Fix and add test cases (introduced by preceding commits)

100pah committed:
Tweak mktest.

100pah committed:
fix: Fix issues in regression test.

100pah committed:
fix: Fix and clarify the definition of axis.`startValue`.

100pah committed:
fix: Fix candlestick clip path on category axis.

100pah committed:
fix_regression: Fix onZero  and startValue backward compatibility.

100pah committed:
fix: (1) Fix: fix bar overflow issue on category axis with boundaryGap: false - use the newly added containShape handler. (2) Fix regression: Drop the previous containShape strategy on bar series, use a simple strategy for better compatibility. (3) Fix regression: Fix new added bar clip feature on single data item case. (4) Fix regression: Clarify code.

100pah committed:
test regression: fix recent broken.

100pah committed:
fix regression: (1) User specified `interval` show be backward compatible. (2) Default IntervalScale extent on single data item case show not be broken. (3) LogScale filter non-positive value.

100pah committed:
fix: (1) Fix time axis reach safeLimit when value scale is milisecond. (2) Fix regression - containShape should be word properly in single value case. And simply the code.

100pah committed:
test: add test.

100pah committed:
fix(progressive): No need to explicitly assign zlevel for el.incremental case.

100pah committed:
fix regression: (1) Revert the recent change to zlevelKey for backward compatibility. (2) Add guard to progressive block.

100pah committed:
fix regression: bandWidth and containShape should properly handle single item and empty.

100pah committed:
fix regression: Fix gauge label precision.

100pah committed:
fix regression: onZero should be backward compatibility. And clarify comments.

100pah committed:
chore: (1) Shift all pure function handlers (legacy pattern) to stage handler to remove unnecessary throw and catch - better for debug; (2) sunburst did not use dataFilter in correct way and it did not work, therefore remove it.

100pah committed:
fix regression: Fix category axis backward compatibility and axis stat travel.

100pah committed:
fix regression: Fix NPE and simplify code.

100pah committed:
test: Add test util for canvas layer debug.

100pah committed:
test: Add visual test marks for v6.1.0.

100pah committed:
Merge branch 'release' into PR/plainheart_fix/alignTicks-precision

100pah committed:
fix regression: fix test cases.

100pah committed:
test:  fix UT.

100pah committed:
Merge branch 'release' into PR/plainheart_fix/alignTicks-precision

100pah committed:
fix: Separate `npm prepare` from other build scripts. The reason is: `npm prepare` is executed when npm install locally or from git, and build/build.js imports a newer version of rollup-terser, which requires newer versions of Node.js (>19). That constraints should not be imposed to users, otherwise some existing pipeline may be broken.

100pah committed:
fix: (1) Fix geo roam (pan and zoom) causes visual artifacts -- geo coord sys and its content (e.g., scatter) should visually align during roaming animation and manual roam. (2) Refactor to uniform roaming impl (via RoamController and ViewCoordSys) and clarify `updateTransform` and related code. (3) Enable scatter and effectScatter clip in geo. (4) Fix aria: palette is modified to ec instance local rather than global, which is more reasonable. (5) Fix map series roaming bug: roam and then click legend hide the first series and then roam and then click legend to restore the first series, the map should not jump.

100pah committed:
dep: rollback @rollup/plugin-terser to 0.4.4 .

100pah committed:
fix: Enable axis statistics to be able to register -- exclude if unnecessary for code size.

100pah committed:
test: Enable echarts-examples/e2e test result to be executed in echarts visual test as supplementary test cases.

100pah committed:
dep: upgrade zrender to 6.1.0 and update package-lock.json.
